processor/spanmetricsprocessor/processor.go

```diff
@@ -141,15 +193,187 @@ func (p *processorImp) ConsumeTraces(ctx context.Context, traces pdata.Traces) e
 // dimensions the user has configured.
 func (p *processorImp) aggregateMetrics(traces pdata.Traces) {
-	// TODO: Add implementation
+	for i := 0; i < traces.ResourceSpans().Len(); i++ {
+		rspans := traces.ResourceSpans().At(i)
+		r := rspans.Resource()
+
+		attr, ok := r.Attributes().Get(conventions.AttributeServiceName)
+		if !ok {
+			continue
+		}
+		serviceName := attr.StringVal()
+		p.aggregateMetricsForServiceSpans(rspans, serviceName)
+	}
+}
+
+func (p *processorImp) aggregateMetricsForServiceSpans(rspans pdata.ResourceSpans, serviceName string) {
+	ilsSlice := rspans.InstrumentationLibrarySpans()
+	for j := 0; j < ilsSlice.Len(); j++ {
+		ils := ilsSlice.At(j)
+		spans := ils.Spans()
+		for k := 0; k < spans.Len(); k++ {
+			span := spans.At(k)
+			p.aggregateMetricsForSpan(serviceName, span)
+		}
+	}
+}
+
+func (p *processorImp) aggregateMetricsForSpan(serviceName string, span pdata.Span) {
+	latencyInMilliseconds := float64(span.EndTime()-span.StartTime()) / float64(time.Millisecond.Nanoseconds())
+
+	// Binary search to find the latencyInMilliseconds bucket index.
+	index := sort.SearchFloat64s(p.latencyBounds, latencyInMilliseconds)
+
+	p.lock.Lock()
+	key := buildKey(serviceName, span, p.dimensions)
+	p.cache(serviceName, span, key)
+	p.updateCallMetrics(key)
+	p.updateLatencyMetrics(key, latencyInMilliseconds, index)
+	p.lock.Unlock()
+}
+
+// updateCallMetrics increments the call count for the given metric key.
+func (p *processorImp) updateCallMetrics(key metricKey) {
+	p.callSum[key]++
+}
+
+// updateLatencyMetrics increments the histogram counts for the given metric key and bucket index.
+func (p *processorImp) updateLatencyMetrics(key metricKey, latency float64, index int) {
+	if _, ok := p.latencyBucketCounts[key]; !ok {
+		p.latencyBucketCounts[key] = make([]uint64, len(p.latencyBounds))
+	}
+	p.latencySum[key] += latency
+	p.latencyCount[key]++
+	p.latencyBucketCounts[key][index]++
+}
+
+func buildDimensionKVs(serviceName string, span pdata.Span, optionalDims []Dimension) dimKV {
+	dims := make(dimKV)
+	dims[serviceNameKey] = serviceName
+	dims[operationKey] = span.Name()
+	dims[spanKindKey] = span.Kind().String()
+	dims[statusCodeKey] = span.Status().Code().String()
+	spanAttr := span.Attributes()
+	var value string
+	for _, d := range optionalDims {
+		// Set the default if configured, otherwise this metric will have no value set for the dimension.
+		if d.Default != nil {
+			value = *d.Default
+		}
+		if attr, ok := spanAttr.Get(d.Name); ok {
+			value = tracetranslator.AttributeValueToString(attr, false)
+		}
+		dims[d.Name] = value
+	}
+	return dims
+}
+
+func concatDimensionValue(metricKeyBuilder *strings.Builder, value string, prefixSep bool) {
+	// It's worth noting that from pprof benchmarks, WriteString is the most expensive operation of this processor.
+	// Specifically, the need to grow the underlying []byte slice to make room for the appended string.
```

Comment:
Nit: we probably can easily have a fairly good idea about the expected size of the final string and can preallocate it. Can be a future improvement.

Reply:
I did try `stringBuilder.Grow(n)` to initialize the capacity before concatenating strings, but it seemed to increase CPU time. It seems the larger proportion of time taken in `Grow` is the creation of a new slice, whereas growing the existing slice seemed to be cheaper; I don't completely understand why tbh.

I've attached screenshots of the flamegraphs in case you were interested.

![Screen Shot 2021-01-29 at 12 15 33 pm](https://user-images.githubusercontent.com/26584478/106218628-c8fbe800-622b-11eb-9965-0204eb5546fe.png)
![Screen Shot 2021-01-29 at 12 14 57 pm](https://user-images.githubusercontent.com/26584478/106218636-cd280580-622b-11eb-8d2f-8b27c5080527.png)